flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableOutputFormat.java

```diff
@@ -0,0 +1,410 @@
+		for (int i = 0; i < rowTypeInfo.getArity() - partitionCols.size(); i++) {
+			objectInspectors.add(HiveTableUtil.getObjectInspector(rowTypeInfo.getTypeAt(i)));
+		}
+
+		if (!isPartitioned) {
+			rowObjectInspector = ObjectInspectorFactory.getStandardStructObjectInspector(
+				Arrays.asList(rowTypeInfo.getFieldNames()),
+				objectInspectors);
+		} else {
+			rowObjectInspector = ObjectInspectorFactory.getStandardStructObjectInspector(
+				Arrays.asList(rowTypeInfo.getFieldNames()).subList(0, rowTypeInfo.getArity() - partitionCols.size()),
+				objectInspectors);
+		}
+	}
+
+	@Override
+	public void writeRecord(Row record) throws IOException {
+		try {
+			HivePartitionWriter partitionWriter = staticWriter;
+			if (isDynamicPartition) {
+				// TODO: to be implemented
+			}
+			partitionWriter.recordWriter.write(serializer.serialize(getConvertedRow(record), rowObjectInspector));
+		} catch (IOException | SerDeException e) {
+			throw new IOException("Could not write Record.", e);
+		}
+	}
+
+	// moves all files under srcDir into destDir
+	private void moveFiles(Path srcDir, Path destDir) throws IOException {
+		if (!srcDir.equals(destDir)) {
+			// TODO: src and dest may be on different FS
+			FileSystem fs = destDir.getFileSystem(jobConf);
+			Preconditions.checkState(fs.exists(destDir) || fs.mkdirs(destDir), "Failed to create dest path " + destDir);
+			if (overwrite) {
+				// delete existing files for overwrite
+				// TODO: support setting auto-purge?
+				final boolean purge = true;
+				// Note we assume the srcDir is a hidden dir, otherwise it will be deleted if it's a sub-dir of destDir
+				FileStatus[] existingFiles = fs.listStatus(destDir, FileUtils.HIDDEN_FILES_PATH_FILTER);
+				if (existingFiles != null) {
+					for (FileStatus existingFile : existingFiles) {
+						Preconditions.checkState(FileUtils.moveToTrash(fs, existingFile.getPath(), jobConf, purge),
+							"Failed to overwrite existing file " + existingFile);
+					}
+				}
+			}
+			FileStatus[] srcFiles = fs.listStatus(srcDir, FileUtils.HIDDEN_FILES_PATH_FILTER);
+			for (FileStatus srcFile : srcFiles) {
+				Path srcPath = srcFile.getPath();
+				Path destPath = new Path(destDir, srcPath.getName());
+				int count = 1;
+				while (!fs.rename(srcPath, destPath)) {
+					String name = srcPath.getName() + "_copy_" + count;
+					destPath = new Path(destDir, name);
+					count++;
+				}
+			}
+		}
+	}
+
+	private void commitJob(String location) throws IOException {
+		jobConf.set(OUTDIR, location);
+		JobContext jobContext = new JobContextImpl(this.jobConf, new JobID());
+		OutputCommitter outputCommitter = this.jobConf.getOutputCommitter();
+		// finalize HDFS output format
+		outputCommitter.commitJob(jobContext);
+	}
+
+	// converts a Row to a list so that Hive can serialize it
+	private Object getConvertedRow(Row record) {
+		int actualArity = isPartitioned ? record.getArity() - partitionCols.size() : record.getArity();
```

Comment:
is the arity the same in the lifecycle of a `HiveTableOutputFormat` instance or can it vary? If the same, shall we make it a member variable and not compute for every record?

Reply:
Makes sense. PR updated.